File: android/src/main/kotlin/info/danielmartinez/fcontacts/FContact.kt
package info.danielmartinez.fcontacts

class FContact (
    val identifier: String,
    var displayName : String? = null,
    var contactType : String? = null,
    var namePrefix : String? = null,
    var givenName : String? = null,
    var middleName : String? = null,
    var familyName : String? = null,
    var nameSuffix : String? = null,
    var nickname : String? = null,
    var jobTitle : String? = null,
    var organizationName : String? = null,
    var birthdayDay : Int? = null,
    var birthdayMonth : Int? = null,
    var birthdayYear : Int? = null,
    var note : String? = null,
    var imageData : ByteArray? = null,
    var thumbnailData : ByteArray? = null,
    var imageDataAvailable : Boolean = false,
    var dates : List<FContactDateLabeled>? = null,  // TODO
    var postalAddresses : List<FContactPostalAddressLabeled>? = null,       // TODO
    var emails : List<FContactValueLabeled>? = null,
    var urls : List<FContactValueLabeled>? = null,
    var phoneNumbers : List<FContactValueLabeled>? = null,
    var socialProfiles : List<FContactSocialProfileLabeled>? = null,
    var contactRelations : List<FContactValueLabeled>? = null,
    var instantMessageAddresses : List<FContactInstantMessageAddressLabeled>? = null    // TODO
) : Comparable<FContact> {

    override fun compareTo( other: FContact ): Int {
        return other.identifier.compareTo( this.identifier )
    }

    fun toMap() : Map<String,Any> {
        val map = mutableMapOf<String,Any>(
            "identifier" to this.identifier,
            "displayName" to if (this.displayName != null) this.displayName as String else ""
        )
        if (this.contactType != null) {
            map["contactType"] = this.contactType as String
        }
        if (this.namePrefix != null) {
            map["namePrefix"] = this.namePrefix as String
        }
        if (this.givenName != null) {
            map["givenName"] = this.givenName as String
        }
        if (this.middleName != null) {
            map["middleName"] = this.middleName as String
        }
        if (this.familyName != null) {
            map["familyName"] = this.familyName as String
        }
        if (this.nameSuffix != null) {
            map["nameSuffix"] = this.nameSuffix as String
        }
        if (this.jobTitle != null) {
            map["jobTitle"] = this.jobTitle as String
        }
        if (this.organizationName != null) {
            map["organizationName"] = this.organizationName as String
        }
        if (this.birthdayDay != null) {
            map["birthdayDay"] = this.birthdayDay as Int
        }
        if (this.birthdayMonth != null) {
            map["birthdayMonth"] = this.birthdayMonth as Int
        }
        if (this.birthdayYear != null) {
            map["birthdayYear"] = this.birthdayYear as Int
        }
        if (this.nickname != null) {
            map["nickname"] = this.nickname as String
        }
        if (this.note != null) {
            map["note"] = this.note as String
        }
        if (this.imageDataAvailable) {
            if (this.imageData != null) {
                map["imageData"] = this.imageData as ByteArray
            }
            if (this.thumbnailData != null) {
                map["thumbnailData"] = this.thumbnailData as ByteArray
            }
        }
        if ((this.emails != null) && (this.emails!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.emails!!.forEach { email ->
                list.add( email.toMap() )
            }
            map["emails"] = list
        }
        if ((this.phoneNumbers != null) && (this.phoneNumbers!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.phoneNumbers!!.forEach { phoneNumber ->
                list.add( phoneNumber.toMap() )
            }
            map["phoneNumbers"] = list
        }
        if ((this.postalAddresses != null) && (this.postalAddresses!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.postalAddresses!!.forEach { postalAddress ->
                list.add( postalAddress.toMap() )
            }
            map["postalAddresses"] = list
        }
        if ((this.dates != null) && (this.dates!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.dates!!.forEach { date ->
                list.add( date.toMap() )
            }
            map["dates"] = list
        }
        if ((this.urls != null) && (this.urls!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.urls!!.forEach { url ->
                list.add( url.toMap() )
            }
            map["urls"] = list
        }
        if ((this.socialProfiles != null) && (this.socialProfiles!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.socialProfiles!!.forEach { socialProfile ->
                list.add( socialProfile.toMap() )
            }
            map["socialProfiles"] = list
        }
        if ((this.contactRelations != null) && (this.contactRelations!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.contactRelations!!.forEach { contactRelation ->
                list.add( contactRelation.toMap() )
            }
            map["contactRelations"] = list
        }
        if ((this.instantMessageAddresses != null) && (this.instantMessageAddresses!!.isNotEmpty())) {
            val list = ArrayList<Map<String,Any>>()
            this.instantMessageAddresses!!.forEach { instantMessageAddress ->
                list.add( instantMessageAddress.toMap() )
            }
            map["instantMessageAddresses"] = list
        }
        return map
    }

}

class FContactValueLabeled (
    private val label: String,
    var value : String? = null
) {
    fun toMap() : Map<String,Any> {
        return mapOf( Pair(this.label, this.value as String) )
    }
}

class FContactDateLabeled (
    private val label: String,
    var day : Int? = null,
    var month : Int? = null,
    var year : Int? = null
) {
    fun toMap() : Map<String,Any> {
        val map = mutableMapOf<String,Any>(
            Pair( "label", this.label )
        )
        if (this.day != null) {
            map[ "day" ] = this.day as Int
        }
        if (this.month != null) {
            map[ "month" ] = this.month as Int
        }
        if (this.year != null) {
            map[ "year" ] = this.year as Int
        }
        return map
    }
}

class FContactPostalAddressLabeled (
    private val label: String,
    var street : String? = null,
    var city : String? = null,
    var neighborhood : String? = null,
    var pobox : String? = null,
    var region : String? = null,
    var postcode : String? = null,
    var country : String? = null,
    var formatted : String? = null
) {
    fun toMap() : Map<String,Any> {
        val map = mutableMapOf<String,Any>(
                Pair( "label", this.label )
        )
        if (this.street != null) {
            map[ "street" ] = this.street as String
        }
        if (this.city != null) {
            map[ "city" ] = this.city as String
        }
        if (this.neighborhood != null) {
            map[ "subLocality" ] = this.neighborhood as String
        }
        if (this.pobox != null) {
            map[ "subAdministrativeArea" ] = this.pobox as String
        }
        if (this.region != null) {
            map[ "state" ] = this.region as String
        }
        if (this.postcode != null) {
            map[ "postalCode" ] = this.postcode as String
        }
        if (this.country != null) {
            map[ "country" ] = this.country as String
        }
        if (this.formatted != null) {
            map[ "formatted" ] = this.formatted as String
        }
        return map
    }
}

class FContactSocialProfileLabeled (
        private val label: String,
        var service : String? = null,
        var userIdentifier : String? = null,
        var username : String? = null,
        var url : String? = null
) {
    fun toMap() : Map<String,Any> {
        val map = mutableMapOf<String,Any>(
                Pair( "label", this.label )
        )
        if (this.service != null) {
            map[ "service" ] = this.service as String
        }
        if (this.userIdentifier != null) {
            map[ "userIdentifier" ] = this.userIdentifier as String
        }
        if (this.username != null) {
            map[ "username" ] = this.username as String
        }
        if (this.url != null) {
            map[ "url" ] = this.url as String
        }
        return map
    }
}

class FContactInstantMessageAddressLabeled (
        private val label: String,
        var service : String? = null,
        var username : String? = null
) {
    fun toMap() : Map<String,Any> {
        val map = mutableMapOf<String,Any>(
            Pair( "label", this.label )
        )
        if (this.service != null) {
            map[ "service" ] = this.service as String
        }
        if (this.username != null) {
            map[ "username" ] = this.username as String
        }
        return map
    }
}

File: ios/Classes/FContact.swift
//
//  FContact.swift
//
//
//  Created by Daniel Martínez on 03/10/2019.
//

import Foundation
import Flutter

public class FContact : Comparable {

    var identifier : String = ""
    var displayName : String = ""
    var contactType : String?
    var namePrefix : String?
    var givenName : String?
    var middleName : String?
    var familyName : String?
    var previousFamilyName : String?
    var nameSuffix : String?
    var nickname : String?
    var phoneticGivenName : String?
    var phoneticMiddleName : String?
    var phoneticFamilyName : String?
    var jobTitle : String?
    var departmentName : String?
    var organizationName : String?
    var phoneticOrganizationName : String?
    var note : String?
    var imageData : Data?
    var thumbnailImageData : Data?
    var imageDataAvailable : Bool = false

    var dates = [FContactDateLabeled]()
    var postalAddresses = [FContactPostalAddressLabeled]()
    var emails = [FContactValueLabeled]()
    var urls = [FContactValueLabeled]()
    var phoneNumbers = [FContactValueLabeled]()
    var socialProfiles = [FContactSocialProfileLabeled]()
    var contactRelations = [FContactValueLabeled]()
    var instantMessageAddresses = [FContactInstantMessageAddressLabeled]()

    public func toMap() -> [String:Any] {
        var result : [String:Any] = [
            "identifier": self.identifier,
            "displayName": self.displayName
        ]
        if let _contactType = self.contactType {
            result[ "contactType" ] = _contactType
        }
        if let _namePrefix = self.namePrefix {
            result[ "namePrefix" ] = _namePrefix
        }
        if let _givenName = self.givenName {
            result[ "givenName" ] = _givenName
        }
        if let _middleName = self.middleName {
            result[ "middleName" ] = _middleName
        }
        if let _familyName = self.familyName {
            result[ "familyName" ] = _familyName
        }
        if let _previousFamilyName = self.previousFamilyName {
            result[ "previousFamilyName" ] = _previousFamilyName
        }
        if let _nameSuffix = self.nameSuffix {
            result[ "nameSuffix" ] = _nameSuffix
        }
        if let _nickname = self.nickname {
            result[ "nickname" ] = _nickname
        }
        if let _phoneticGivenName = self.phoneticGivenName {
            result[ "phoneticGivenName" ] = _phoneticGivenName
        }
        if let _phoneticMiddleName = self.phoneticMiddleName {
            result[ "phoneticMiddleName" ] = _phoneticMiddleName
        }
        if let _phoneticFamilyName = self.phoneticFamilyName {
            result[ "phoneticFamilyName" ] = _phoneticFamilyName
        }
        if let _jobTitle = self.jobTitle {
            result[ "jobTitle" ] = _jobTitle
        }
        if let _departmentName = self.departmentName {
            result[ "departmentName" ] = _departmentName
        }
        if let _organizationName = self.organizationName {
            result[ "organizationName" ] = _organizationName
        }
        if let _phoneticOrganizationName = self.phoneticOrganizationName {
            result[ "phoneticOrganizationName" ] = _phoneticOrganizationName
        }
        if self.dates.count > 0 {
            var _dates = [[String:Any]]()
            for date in self.dates {
                var _components = [String:Any]()
                _components["label"] = date.label as String
                if let _day = date.valueDay {
                    _components["day"] = _day
                }
                if let _month = date.valueMonth {
                    _components["month"] = _month
                }
                if let _year = date.valueYear {
                    _components["year"] = _year
                }
                _dates.append( _components )
            }
            result[ "dates" ] = _dates
        }
        if let _note = self.note {
            result[ "note" ] = _note
        }
        result[ "imageDataAvailable" ] = self.imageDataAvailable
        if let imageData = self.imageData {
            result[ "imageData" ] = FlutterStandardTypedData( bytes: imageData )
        }
        if let thumbnailImageData = self.thumbnailImageData {
            result[ "thumbnailImageData" ] = FlutterStandardTypedData( bytes: thumbnailImageData )
        }
        if self.postalAddresses.count > 0 {
            var _items = [[String:Any]]()
            for postalAddress in self.postalAddresses {
                var _components = [String:String]()
                _components["label"] = postalAddress.label
                if let _street = postalAddress.valueStreet {
                    _components["street"] = _street
                }
                if let _city = postalAddress.valueCity {
                    _components["city"] = _city
                }
                if let _subLocality = postalAddress.valueSubLocality {
                    _components["subLocality"] = _subLocality
                }
                if let _subAdministrativeArea = postalAddress.valueSubAdministrativeArea {
                    _components["subAdministrativeArea"] = _subAdministrativeArea
                }
                if let _postalCode = postalAddress.valuePostalCode {
                    _components["postalCode"] = _postalCode
                }
                if let _state = postalAddress.valueState {
                    _components["state"] = _state
                }
                if let _country = postalAddress.valueCountry {
                    _components["country"] = _country
                }
                if let _isoCountryCode = postalAddress.valueISOCountryCode {
                    _components["isoCountryCode"] = _isoCountryCode
                }
                if let _formatted = postalAddress.formatted {
                    _components["formatted"] = _formatted
                }
                _items.append( _components )
            }
            result[ "postalAddresses" ] = _items
        }
        if self.emails.count > 0 {
            var _items = [[String:Any]]()
            for email in self.emails {
                _items.append([
                    email.label: email.value
                ])
            }
            result[ "emails" ] = _items
        }
        if self.urls.count > 0 {
            var _items = [[String:Any]]()
            for url in self.urls {
                _items.append([
                    url.label: url.value
                ])
            }
            result[ "urls" ] = _items
        }
        if self.phoneNumbers.count > 0 {
            var _items = [[String:Any]]()
            for phoneNumber in self.phoneNumbers {
                _items.append([
                    phoneNumber.label: phoneNumber.value
                ])
            }
            result[ "phoneNumbers" ] = _items
        }
        if self.socialProfiles.count > 0 {
            var _items = [[String:Any]]()
            for socialProfile in self.socialProfiles {
                var _components = [String:String]()
                _components["label"] = socialProfile.label
                if let _service = socialProfile.valueService {
                    _components["service"] = _service
                }
                if let _userIdentifier = socialProfile.valueUserIdentifier {
                    _components["userIdentifier"] = _userIdentifier
                }
                if let _username = socialProfile.valueUsername {
                    _components["username"] = _username
                }
                if let _url = socialProfile.valueUrl {
                    _components["url"] = _url
                }
                _items.append( _components )
            }
            result[ "socialProfiles" ] = _items
        }
        if self.contactRelations.count > 0 {
            var _items = [[String:Any]]()
            for contactRelation in self.contactRelations {
                _items.append([
                    contactRelation.label: contactRelation.value
                ])
            }
            result[ "contactRelations" ] = _items
        }
        if self.instantMessageAddresses.count > 0 {
            var _items = [[String:Any]]()
            for instantMessageAddress in self.instantMessageAddresses {
                var _components = [String:String]()
                _components["label"] = instantMessageAddress.label
                if let _service = instantMessageAddress.valueService {
                    _components["service"] = _service
                }
                if let _username = instantMessageAddress.valueUsername {
                    _components["username"] = _username
                }
                _items.append( _components )
            }
            result[ "instantMessageAddresses" ] = _items
        }

        return result
    }

    public static func == (lhs: FContact, rhs: FContact) -> Bool {
        var returnValue = false
        if (lhs.identifier == rhs.identifier) {
            returnValue = true
        }
        return returnValue
    }

    public static func < (lhs: FContact, rhs: FContact) -> Bool {
        var returnValue = false
        if (lhs.identifier < rhs.identifier) {
            returnValue = true
        }
        return returnValue
    }

}

public class FContactValueLabeled {
    var label: String = ""
    var value: String = ""
}

public class FContactDateLabeled {
    var label: String = ""
    var valueDay: Int?
    var valueMonth: Int?
    var valueYear: Int?
}

public class FContactPostalAddressLabeled {
    var label: String = ""
    var valueStreet: String?
    var valueCity: String?
    var valueSubLocality: String?
    var valueSubAdministrativeArea: String?
    var valuePostalCode: String?
    var valueState: String?
    var valueCountry: String?
    var valueISOCountryCode: String?
    var formatted: String?
}

public class FContactSocialProfileLabeled {
    var label: String = ""
    var valueService: String?
    var valueUserIdentifier: String?
    var valueUsername: String?
    var valueUrl: String?
}

public class FContactInstantMessageAddressLabeled {
    var label: String = ""
    var valueService: String?
    var valueUsername: String?
}
